src/vm/types/program.zig

```diff
@@ -33,32 +33,116 @@ pub const HintRange = struct {
+            .Extensive => self.Extensive.deinit(),
+            .NonExtensive => self.NonExtensive.deinit(),
+        }
+    }
+};
+
+const Hints = std.ArrayList(HintParams);
+
 /// Represents a collection of hints.
 ///
 /// This structure contains a list of `HintParams` and a map of `HintRange` corresponding to a `Relocatable`.
 pub const HintsCollection = struct {
     const Self = @This();
     /// List of HintParams.
-    hints: std.ArrayList(HintParams),
+    hints: Hints,
     /// Map of Relocatable to HintRange.
-    hints_ranges: std.HashMap(
-        Relocatable,
-        HintRange,
-        std.hash_map.AutoContext(Relocatable),
-        std.hash_map.default_max_load_percentage,
-    ),
+    hints_ranges: HintsRanges,
 
     /// Initializes a new HintsCollection with default values.
     ///
     /// # Params:
     ///   - `allocator`: The allocator used to initialize the collection.
+    pub fn init(allocator: Allocator, hints: std.AutoHashMap(usize, []const HintParams), program_length: usize, extensive_hints: bool) !Self {
+        var max_hint_pc = @as(usize, @intCast(0));
+        var total_hints_len = @as(usize, @intCast(0));
+        var it = hints.iterator();
+        while (it.next()) |kv| {
+            max_hint_pc = @max(max_hint_pc, kv.key_ptr.*);
+            total_hints_len += kv.value_ptr.len;
+        }
+
+        if (max_hint_pc == 0 or total_hints_len == 0) {
+            return Self.initDefault(allocator);
+        }
+
+        if (max_hint_pc >= program_length) {
+            return ProgramError.InvalidHintPc;
+        }
+
+        var hints_values = try std.ArrayList(HintParams).initCapacity(allocator, total_hints_len);
+        var hints_ranges_non_ext: ?std.ArrayList(HintRange) = null;
+        var hints_ranges_ext: ?std.AutoHashMap(Relocatable, HintRange) = null;
+
+        if (extensive_hints) {
+            hints_ranges_ext = std.AutoHashMap(Relocatable, HintRange).init(allocator);
+        } else {
+            hints_ranges_non_ext = std.ArrayList(HintRange).init(allocator);
+        }
+
+        it = hints.iterator();
+        while (it.next()) |kv| {
+            if (kv.value_ptr.*.len > 0) {
+                const range = HintRange{ .start = hints_values.items.len, .length = kv.value_ptr.len };
+                if (extensive_hints) {
+                    try hints_ranges_ext.?.put(Relocatable.init(0, kv.key_ptr.*), range);
+                } else {
+                    try hints_ranges_non_ext.?.append(range);
+                }
+                try hints_values.appendSlice(kv.value_ptr.*);
+            }
+        }
+
+        if (extensive_hints) {
+            return Self{
+                .hints = hints_values,
+                .hints_ranges = HintsRanges{ .Extensive = hints_ranges_ext.? },
+            };
+        } else {
+            return Self{
+                .hints = hints_values,
+                .hints_ranges = HintsRanges{ .NonExtensive = hints_ranges_non_ext.? },
+            };
+        }
```

Comment:
```suggestion
        if (extensive_hints) {
            return .{
                .hints = hints_values,
                .hints_ranges = .{ .Extensive = hints_ranges_ext.? },
            };
        } else {
            return .{
                .hints = hints_values,
                .hints_ranges = .{ .NonExtensive = hints_ranges_non_ext.? },
            };
        }
```

Reply:
This won't be needed after init in HintRanges